### DISCUSSION TITLE
[useLocalStorage] Synchronize changes in local storage and data value

### DISCUSSION THREAD

**Discussion by nooooooom** [vueuse/vueuse#3632] [General]
### Describe the bug

When I set a value to `useLocalStorage`, I cannot read the value through `localStorage.getItem` in real-time.

I think the value of `useLocalStorage` should be derived from `localStorage` as it describes and should not be ahead of `localStorage`. Or at least they should be set simultaneously.

### Reproduction

https://stackblitz.com/edit/vitejs-vite-qjrcxk/?file=src%2Fmain.js,src%2FApp.vue

### System Info

```Shell
none
```

### Used Package Manager

pnpm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vueuse/vueuse/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vueuse.org/guide).
- [X] Check that there isn't [already an issue](https://github.com/vueuse/vueuse/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a VueUse issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to https://github.com/vuejs/core instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vueuse/vueuse/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by Doctor-wu** [comment]
When you set the value of `cache`, Vueuse will not perceive it immediately. Instead, Vueuse will be notified the cache has changed in the nextTick cuz `useLocalStorage` watch the `cache`, that's the behavior of [Vue's Reative System](https://vuejs.org/guide/extras/reactivity-in-depth.html)

**Comment by nooooooom** [comment]
> When you set the value of `cache`, Vueuse will not perceive it immediately. Instead, Vueuse will be notified the cache has changed in the nextTick cuz `useLocalStorage` watch the `cache`, that's the behavior of [Vue's Reative System](https://vuejs.org/guide/extras/reactivity-in-depth.html)

I understand its internal implementation mechanism, but I think its original data flow should always be maintained. The value of `useLocalStorage` should always come from `localStorage` and `defaultValue`.

**Comment by Doctor-wu** [comment]
If you read from the `cache`, it comes from`defaultValue` indeed. In your case, you mix use`localStorage` with `useLocalStorage`. Vueuse doesn't perceive the mutate of `cache`, therefore it doesn't have time to update `localStorage`, that's why you got the old value from `localStorage`.

**Comment by nooooooom** [comment]
> If you read from the `cache`, it comes from`defaultValue` indeed. In your case, you mix use`localStorage` with `useLocalStorage`. Vueuse doesn't perceive the mutate of `cache`, therefore it doesn't have time to update `localStorage`, that's why you got the old value from `localStorage`.

You are right, but this is because the implementation of `useLocalStorage` makes the update of `localStorage` lag behind the data returned by `useLocalStorage`. Throughout the update process, the data flow direction should remain `user input` -> `localStorage change` -> `useLocalStorage data change` -> `user view`.

Just like when the user clicks a button on the page, Vue should update the data first and then update the view based on the changes in data, instead of updating the view first and then synchronizing the data. In this example, `useLocalStorage` is like a view that should always be driven by `localStorage`.

**Comment by Doctor-wu** [comment]
That's not the same concept 😂, I understand your `flow`, but it's different from what we are talking about. If you are using `useLocalStorage`, then you shouldn't get value from`localStorage` directly. You get value from `cache`, and you set value from `cache`, that's a convenient entry with reactivity for you to control `localStorage`, that's it.

**Comment by nooooooom** [comment]
No matter how we update the value of `useLocalStorage`, it should always synchronize with the value of `localStorage`. Consider the following example

```
const cache = useLocalStorage('key', '')
cache.value = 'a'
localStorage.setItem('key', 'b')
```

**Comment by Doctor-wu** [comment]
As I had said above, you are not supposed to use`localStorage` directly if you are using `useLocalStorage`. If we change `localStorage` synchronize and user change `cache.value` lots of times in a single tick which is very possible,  then it will sync `localStorage` lots of times which is redundancy

**Comment by nooooooom** [comment]
> That's not the same concept 😂, I understand your `flow`, but it's different from what we are talking about. If you are using `useLocalStorage`, then you shouldn't get value from`localStorage` directly. You get value from `cache`, and you set value from `cache`, that's a convenient entry with reactivity for you to control `localStorage`, that's it.

On the contrary, the data flow is its core. If a derived state cannot guarantee its data flow direction, then I think its implementation concept should be corrected.

Like this computed:
```js
const number = ref(0)
const double = computed(() => number.value * 2)
```
If I want to set its value, do I need to first write a watch function to observe the changes in computed, and then modify the `number` based on the value of `double`?
```js
watch(double, (double) => {
  value.value = double / 2
})
```
No, we should set the value of `number` and then double will change accordingly
```js
const double = computed({
  get: () => number.value * 2,
  set: (value) => number.value = value / 2
})
```

 The data flow is its core.

**Comment by Doctor-wu** [comment]
Just jump out of that `flow`, `useLocalStorage` just connect you and `localStorage` in a **reactive** way, that's the key, the reactive system will collect the synchronize mutate and update asynchronized to prevent redundancy update, if `useLocalStorage` reactive with each mutate synchronizely, it will cause redundancy update.

**Comment by nooooooom** [comment]
> Just jump out of that `flow`, `useLocalStorage` just connect you and `localStorage` in a **reactive** way, that's the key, the reactive system will collect the synchronize mutate and update asynchronized to prevent redundancy update, if `useLocalStorage` reactive with each mutate synchronizely, it will cause redundancy update.

How the value is set is not something that should be optimized. What needs to be optimized is what happens when the value is set. For example, if I click the button ten times in one frame, then the count value should change from 0 to 10 in sequence, but I It is possible to update the rendered view only once. 

Data flow should not be broken.

**Comment by Doctor-wu** [comment]
Data flow doesn't break if you don't mix using `useLocalStorage` and `localStorage`. `localStorage`'s API are synchronized, but reactive system bring the asynchronize, but it's ok cuz it's self-complete. The real flow should be  `mutate` -> `useLocalStorage` -> `localStorage`, you are not supposed to kick `useLocalStorage` away

**Comment by Doctor-wu** [comment]
<img width="944" alt="CleanShot 2023-12-06 at 17 59 04@2x" src="https://github.com/vueuse/vueuse/assets/44631608/80c20f50-4338-4081-b9d3-e89086fcff59">

**Comment by nooooooom** [comment]
> Data flow doesn't break if you don't mix using `useLocalStorage` and `localStorage`. `localStorage`'s API are synchronized, but reactive system bring the asynchronize, but it's ok cuz it's self-complete. The real flow should be `mutate` -> `useLocalStorage` -> `localStorage`, you are not supposed to kick `useLocalStorage` away

`useLocalStorage` 监听 `localStorage` 的更新，然后赋值给响应式的数据，这是没问题的，并且有且只有这一种数据流向。当向 `useLocalStorage` 设置值的时候， `vueuse` 内部的做法是先更新响应式的数据的 `value`，再通过监听这个数据的改变将最新的值写入 `localStorage`，这个时候问题就出现了，`watch` 普通使用时它不是一个同步进行的东西，也就是说我每次向 `useLocalStorage` 传值的时候，`localStorage` 需要在下一个时间点才能得以更新，但是依赖于 `localStorage` 的 `useLocalStorage`  所返回的那个响应式数据却得到更新了，这是不应该的。

在正常的数据流中，不管在哪个阶段去看 `useLocalStorage` 与  `localStorage` 的值， `localStorage` 都应该是领先或等同于 `useLocalStorage` 的，但是 `useLocalStorage` 却破坏了这个规则，它基于 `localStorage` 却可能领先于 `localStorage`。

上面的图像中有个误区就是 `useLocalStorage` 与 `localStorage` 并不是双向的，`useLocalStorage` 的值只会来源于 `localStorage`，而  `useLocalStorage`  也只是将修改 `localStorage`  的行为封装了起来而已，用户在设置  `useLocalStorage`  的值的时候，实际上是直接对 `localStorage` 进行赋值。

**Comment by nooooooom** [comment]
我觉得这个例子所体现出来的问题已经很明显了，我们可以把这两种赋值行为看作是不同框架或者不同地方对 `localStorage` 的值的修改，但是 `useLocalStorage` 对 `localStorage` 的修改却是异步的。如果 `useLocalStorage` 的设计理念便是如此，那我认为 `useLocalStorage` 所返回的响应式数据的值也应该被异步赋值。

```js
const cache = useLocalStorage('key', '')
cache.value = 'a'
localStorage.setItem('key', 'b')
```

**Comment by Doctor-wu** [comment]
> > Data flow doesn't break if you don't mix using `useLocalStorage` and `localStorage`. `localStorage`'s API are synchronized, but reactive system bring the asynchronize, but it's ok cuz it's self-complete. The real flow should be `mutate` -> `useLocalStorage` -> `localStorage`, you are not supposed to kick `useLocalStorage` away
> 
> `useLocalStorage` 监听 `localStorage` 的更新，然后赋值给响应式的数据，这是没问题的，并且有且只有这一种数据流向。当向 `useLocalStorage` 设置值的时候， `vueuse` 内部的做法是先更新响应式的数据的 `value`，再通过监听这个数据的改变将最新的值写入 `localStorage`，这个时候问题就出现了，`watch` 普通使用时它不是一个同步进行的东西，也就是说我每次向 `useLocalStorage` 传值的时候，`localStorage` 需要在下一个时间点才能得以更新，但是依赖于 `localStorage` 的 `useLocalStorage` 所返回的那个响应式数据却得到更新了，这是不应该的。
> 
> 在正常的数据流中，不管在哪个阶段去看 `useLocalStorage` 与 `localStorage` 的值， `localStorage` 都应该是领先或等同于 `useLocalStorage` 的，但是 `useLocalStorage` 却破坏了这个规则，它基于 `localStorage` 却可能领先于 `localStorage`。
> 
> 上面的图像中有个误区就是 `useLocalStorage` 与 `localStorage` 并不是双向的，`useLocalStorage` 的值只会来源于 `localStorage`，而 `useLocalStorage` 也只是将修改 `localStorage` 的行为封装了起来而已，用户在设置 `useLocalStorage` 的值的时候，实际上是直接对 `localStorage` 进行赋值。

It's not mutate internal

**Comment by Doctor-wu** [comment]
That's the same scene when you `watch` the `cache` to subscribe the mutate of storage. When the storage event received by `useLocalStorage`, it will update the value of `cache` and you will perceive the mutate in nextTick.

**Comment by nooooooom** [comment]
> That's the same scene when you `watch` the `cache` to subscribe the mutate of storage. When the storage event received by `useLocalStorage`, it will update the value of `cache` and you will perceive the mutate in nextTick.

I think we should admit the error caused by this approach (the value of `useLocalStorage` is earlier than the value of `localStorage`) and eliminate it, rather than let the user understand why there are unexpected results based on this implementation.

The value of `useLocalStorage` comes from `localStorage`, but it is assigned earlier than `localStorage`. This is a logical misunderstanding. Although we can try to eliminate the problems caused by asynchrony by passing in `{ flush: 'sync' }`, it still violates the principle of assignment.

I could give many examples of the problems this implementation creates. For example, I hacked the implementation of `localStorage.setItem` and bound a middleware to it to prevent the storage of values that do not meet the business specifications. In this case, even if I pass in `{ flush: 'sync' }` to "try to" synchronize the timing of the assignment, when the behavior of `localStorage.setItem` is blocked, `useLocalStorage` cannot restore to the previous value.

Data flow is at its core, and you can't use Vue's reactive mechanism to destroy it and explain to users that "it's Vue that makes it impossible for us to guarantee timing."

**Comment by Doctor-wu** [comment]
Hi @nooooooom , I would like to transfer this issue into a discussion, looking forward to more ideas from the community~

**Comment by MaxMooc** [comment]
有结果了吗？是不是还没解决？用的原生api

**Comment by ilyaliao** [comment]
I want to close this issue as it goes beyond the responsibility of `vueuse`.

If you want to read a value from `localStorage`, you can use the following structure.

```
nextTick(async () => {
  cache.value = 'next';
  nextTick(() => {
    syncValue.value = localStorage.getItem('cache_value');
    nextTick(() => {
      nextTickValue.value = localStorage.getItem('cache_value');
    });
  })
});
```
